### PR TITLE
`Development`: Remove unnecessary path parameters

### PR DESCRIPTION
**Pull request by MarkusPaulsen**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context
The FileResource class unnecessarily uses some additional path parameters, which reduces simplicity and increases the chances of bugs and/or missuse.

### Description
Removed the aforementioned parameters from the paths.

### Steps for Testing
We just want to make sure that I didn't accidentally break the implementation.

Prerequisites:
- 1 Instructor
- 1 Student

Course (Instructor/Student):
1. Look at a course with course image (e.g. 6.0 Testing Session Course 1 on TS1). Alternatively, create one
2. Make sure the image gets rendered correctly

After each step, commit and see if the changes were persisted by reloading
Programming Exercise (Instructor, Student):
1. Create a programming exercise with the online editor activated.
2. Change the programming language and make sure the template problem statement gets changed.
3. Actually create the exercise (submit) and make sure all repos got created correctly (tests should return correctly for solution and template.
4. Feel free to check a simple student submission too
5. No issues should occur

Fileupload Exercise (Instructor, Student):
1. Create a file upload exercise with a description. Upload an image to the description
2. Start the exercise, the image should show up properly as well as the rest of the description
3. Participate in the exercise
4. Upload another file
5. View/Download your submission
6. No issues should occur, the correct file should download.

Quiz Exercise (Instructor):
1. Create a Quiz Exercise with a Drag and Drop Question including Background and image drag items
2. Preview the quiz and make sure all images get rendered properly

#### Exam Mode Testing

We would need to test the participation checker which requires testing it on an app. However, the changes to the two endpoints are the exact same changes I did to a few other endpoints. Meaning if one fails, all should fail.
=> I deem it unnecessary to test every single of these endpoints. They already get tested by the other endpoints.

#### Performance Review
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

I still added the coverage here, but there were no functional changes only some small clean-up which is covered, so I did not add more tests for things I didn't touch in this PR.
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| FileResource.java | 70.2% |                            |
| programming-exercise-update.component.ts | 79.9% |               |
| code-editor-file-browser.component.ts | 85.4% |               |
| file.service.ts | 47.3% |               |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/DragItemRepository.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/DragItemRepository.java:30]:
```suggestion
    default DragItem findDragItemByIdWithEagerQuestionAndExerciseAndCourseOrThrow(Long dragItemId) {
```

File: src/main/java/de/tum/in/www1/artemis/repository/DragItemRepository.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/DragItemRepository.java:31]:
```suggestion
        return findDragItemByIdWithEagerQuestionAndExerciseAndCourse(dragItemId).orElseThrow(() -> new EntityNotFoundException("DragItem", dragItemId));
```

File: src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java:39]:
```suggestion
    default DragAndDropQuestion findDragAndDropQuizQuestionByIdWithEagerExerciseAndCourseOrThrow(Long questionId) {
```

File: src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/repository/QuizQuestionRepository.java:40]:
```suggestion
        return findDragAndDropQuizQuestionByIdWithEagerExerciseAndCourse(questionId).orElseThrow(() -> new EntityNotFoundException("DragAndDropQuestion", questionId));
```

**Review by MaximilianAnzinger — COMMENTED** [review]
Left some suggestions.

File: src/main/java/de/tum/in/www1/artemis/repository/DragItemRepository.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/DragItemRepository.java:24]: some fetches are not necessary as they would automatically be fetched by Hibernate due to ManyToOne being Eager

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:1]: Is it necessary to map all suffixes for the end points `"files/drag-and-drop/backgrounds/{questionId}/*"` and `"files/drag-and-drop/drag-items/{dragItemId}/*"`.
For me, that's not a major concern, but this might lead to confusion in the future.

**Review by MaximilianAnzinger — APPROVED** [review]
Thank you for addressing my suggestions. I left one comment/question 👍

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:1]: We need to map something; otherwise, the path won't get recognised. I think, in the future, it would be wise to refactor such paths in general to maybe something like `/api/courses/{courseId}/quiz-exercises/{exerciseId}/questions/{questionId}/background` or whatever. Then we don't even need those conversions between public and private paths anymore and also don't have to retrieve any path.

But that's just vague ideas.

**Review by RY997 — DISMISSED** [review]
Tested on ts3, worked well

**Review by tobias-lippert — APPROVED** [review]
Tested quiz exercises in testing session. Everything still worked.

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS3. Works according to testing steps besides the part with the "natural language". I'm confused about what I should test here. The language of the UI changes as expected and I don't think there are multiple templates for multiple languages.

**Review by DominikRemo — DISMISSED** [review]
Tested in testing session on ts3 for quiz exercises

**Review by JohannesStoehr — COMMENTED** [review]
Tested on TS3 during testing session. Programming exercises work as expected

**Comment by julian-christl** [comment]
> Tested on TS3. Works according to testing steps besides the part with the "natural language". I'm confused about what I should test here. The language of the UI changes as expected and I don't think there are multiple templates for multiple languages.

@MaximilianAnzinger That was me being stupid: There are some templates in English and some in German.

**Review by JohannesStoehr — APPROVED** [review]

File: src/main/webapp/app/shared/http/file.service.ts
Comment (Strohgelaender) [inline src/main/webapp/app/shared/http/file.service.ts:18]: Why is language optional here? On the server the language must be present.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:227]: No, then the mapping wouldn't be correct.

File: src/main/webapp/app/shared/http/file.service.ts
Comment (julian-christl) [reply src/main/webapp/app/shared/http/file.service.ts:18]: There is a README.md in the root of the templates.
As this PR doesn't intend to modify the existing behaviour, I left it like that. If someone want's to remove this because it's unused, I have no objections (other than please do it after this PR is merged xD)

Comment (Strohgelaender) [reply src/main/webapp/app/shared/http/file.service.ts:18]: ` public ResponseEntity<byte[]> getTemplateFile(@PathVariable ProgrammingLanguage language, @PathVariable Optional<ProjectType> projectType)`

If there is a root-readme it's unaccessible since the server endpoint does to allow the Programming language to be missing.

**Review by nityanandaz — APPROVED** [review]
Tested on TS3 during testing session (2023-08-15), file upload exercises are working

File: src/main/webapp/app/shared/http/file.service.ts
Comment (julian-christl) [reply src/main/webapp/app/shared/http/file.service.ts:18]: True, I can actually adapt this in the client.

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:287]: Barely any, I feel, but it's beyond the scope of this PR. I only adapted the parameters slightly so I'd like to avoid enlarging the scope here.

**Review by Strohgelaender — APPROVED** [review]
lgtm, thanks for adressing my feedback.

**Review by MaximilianAnzinger — APPROVED** [review]
Reapproved code

**Review by RY997 — APPROVED** [review]
Reapprove!

**Review by JohannesStoehr — APPROVED** [review]
Reapprove

**Review by nityanandaz — APPROVED** [review]
Manuel test on TS1 (2023-09-04); iOS app still shows assets

**Review by dearjasmina — APPROVED** [review]
tested and works as expected, code looks good 👍🏻

**Review by jakubriegel — APPROVED** [review]
Tested live with Markus

**Review by MarkusPaulsen — COMMENTED** [review]
Maintainer-Approved
- Code looks good
- Tested on TS1

**Pull request merged**
